src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,481 @@
+    """An int option."""
+
+    option_type: Any = int
+
+    @overload
+    def __new__(cls, *flag_names: str, default: int, help: str) -> IntOption[int]:
+        ...
+
+    @overload
+    def __new__(cls, *flag_names: str, help: str) -> IntOption[int | None]:
+        ...
+
+    def __new__(cls, *flag_names, default=None, help):
+        return super().__new__(cls, *flag_names, default=default, help=help)
+
+
+class FloatOption(OptionBase[FloatT]):
+    """A float option."""
+
+    option_type: Any = float
+
+    @overload
+    def __new__(cls, *flag_names: str, default: float, help: str) -> FloatOption[float]:
+        ...
+
+    @overload
+    def __new__(cls, *flag_names: str, help: str) -> FloatOption[float | None]:
+        ...
+
+    def __new__(cls, *flag_names, default=None, help):
+        return super().__new__(cls, *flag_names, default=default, help=help)
+
+
+class BoolOption(OptionBase[BoolT]):
+    """A bool option.
+
+    If you don't provide a `default` value, this becomes a "tri-bool" where the property will return
+    `None` if unset by the user.
+    """
+
+    option_type: Any = bool
+
+    @overload
+    def __new__(cls, *flag_names: str, default: bool, help: str) -> BoolOption[bool]:
+        ...
+
+    @overload
+    def __new__(cls, *flag_names: str, help: str) -> BoolOption[bool | None]:
+        ...
+
+    def __new__(cls, *flag_names, default=None, help):
+        return super().__new__(cls, *flag_names, default=default, help=help)
+
+
+class EnumOption(OptionBase[PropType], Generic[PropType]):
+    """An Enum option.
+
+    If you provide a `default` parameter, the `option_type` parameter will be inferred from the type
+    of the default. Otherwise, you'll need to provide the `option_type`.
+    In either case, mypy will infer the correct Generic's type-parameter, so you shouldn't need to
+    provide it.
+
+    E.g.
+        EnumOption(..., option_type=MyEnum)  # property type is deduced as `MyEnum | None`
+        EnumOption(..., default=MyEnum.Value)  # property type is deduced as `MyEnum`
+    """
+
+    @overload
+    def __new__(cls, *flag_names: str, default: EnumT, help: str) -> EnumOption[EnumT]:
+        ...
+
+    # N.B. This has an additional param for the no-default-provided case: `option_type`.
+    @overload
+    def __new__(
+        cls, *flag_names: str, option_type: Type[EnumT], help: str
+    ) -> EnumOption[EnumT | None]:
+        ...
+
+    def __new__(
+        cls,
+        *flag_names,
+        option_type=None,
+        default=None,
+        help,
+    ):
+        if option_type is None:
+            if default is None:
+                raise ValueError("`option_type` must be provided if `default` isn't provided.")
+            option_type = type(default)
+
+        return super().__new__(
+            cls, *flag_names, default=default, option_type=option_type, help=help
+        )
+
+
+class DictOption(OptionBase[FrozenDict[str, ValueType]], Generic[ValueType]):
```

Comment:
I expect this might give us problems because `FrozenDict` requires both keys and values to be hashable. You'll have issues if the values are lists, for example.

But let's land this as is, and then we can improve it if we find we need to.

Reply:
Can't we rely on/enforce tuples in place of lists?

Reply:
Is that mechanically enforced, or a gentlemen's agreement?
I can test

Reply:
Yeah, but we have to eagerly coerce the `.values()` into tuples. 

https://github.com/pantsbuild/pants/blob/d0eb5be8e13b41fe3146ad14293facbd88156742/src/python/pants/engine/target.py#L1498-L1525

To do that, we need to know you want us to do that. It would be wrong to call `tuple()` on a `DictStringToStringOption`.

I think we can/should punt on this all, though. This PR is awesome as is.

One thing we could consider is not adding `DictOption` in this PR? If you want to use that, you'll have to still use the old API until we have time to fully flesh this out. Then we don't risk breaking anyone using this (experimental) feature.

Reply:
I guess my question is if people use `FrozenDict[str, list[int]]` would that pass `mypy`.

Reply:
Let's punt on adding Dict and revisit. Git history will tell us how to add it back in.